caikit/core/toolkit/destroyable_process.py

```diff
@@ -0,0 +1,193 @@
+
+    @property
+    def canceled(self) -> bool:
+        return self.__canceled
+
+    @property
+    def ran(self) -> bool:
+        return self.__started and not self.is_alive()
+
+    @property
+    def threw(self) -> bool:
+        return self.error is not None
+
+    def get_or_throw(self) -> Any:
+        """Get the result of the execution or raise an error if one occurred"""
+        if self.destroyed:
+            log.error(
+                "<COR24981767E>",
+                "get_or_throw called on destroyed process, no value to return",
+            )
+
+        if not self.ran:
+            log.error(
+                "<COR12037430E>",
+                "get_or_throw called on process, but it has not finished running",
+            )
+
+        # Update the result and throw if it's an error
+        error = self.error
+        if error is not None:
+            raise error
+        return self.__result
+
+    def destroy(self):
+        """Cancel any in-progress work"""
+        self.__destroyed = True
+        if self.is_alive() or not self.__started:
+            self.__canceled = True
+        if self.__started:
+            self.terminate()
+            self.join(self._destroy_grace_period)
+            self.kill()
+            self.join()
+            self._completion_event.set()
+
+    ## Process Interface ##
+
+    def start(self):
+        if self.destroyed:
+            err_msg = "Not starting work on pre-canceled process"
+            log.warning("<COR42191929W>", err_msg)
+            self._completion_event.set()
+            self.__result = RuntimeError(err_msg)
+            return
+        self.__started = True
+        return super().start()
+
+    def join(self, *args, **kwargs):
+        if self.destroyed and not self.__started:
+            return
+        return super().join(*args, **kwargs)
+
+    # NOTE: This functionality is covered by the unit tests, but pytest-cov is
+    #   not correctly collecting the coverage info since it executes inside the
+    #   subprocess
+    def run(self):  # pragma: no cover
+        try:
+            # Run and indicate to the parent that no
+            super().run()
+
+        # Catch any errors thrown within a subprocess so that they can be
+        # forwarded to the parent
+        # pylint: disable=broad-exception-caught
+        except Exception as err:
+            err_str = repr(err)
+            log.error(
+                "<COR69863806E>",
+                "Caught exception in destroyable process: %s",
+                err_str,
+                exc_info=True,
+            )
+            self._child_conn.send(err)
+        finally:
+            self._completion_event.set()
+
+    ## Impl ##
+
+    def _update_result(self):
+        if self._parent_conn.poll():
+            self.__result = self._parent_conn.recv()
+
+    @property
+    def error(self) -> Optional[Exception]:
+        self._update_result()
+        if isinstance(self.__result, Exception):
```

Comment:
hmm. `self.__result` will not be an instance of Exception since it will happen in the subprocess right?